File: src/model/session/session.rs
use crate::model::session::traits::*;
use crate::model::session::status::Status;
use crate::model::client::traits::Controller as ClientController;

use std::time::{Duration, Instant};

//use diesel;
//use diesel::prelude::*;
//use diesel::mysql::MysqlConnection;
//
//#[derive(Queryable)]
pub struct Session {
    pub id: i32,
    pub cookie: String,
    pub created_at: Instant,
    pub touch_at: Instant,
    pub timeout: Duration,
    pub status: Status,
    client: Box<dyn ClientController>,
}

impl Session {
    pub fn new(client: Box<dyn ClientController>, cookie: String, timeout: Duration) -> Self {
        Session{
            id: 0,
            cookie: cookie,
            created_at: Instant::now(),
            touch_at: Instant::now(),
            timeout: timeout,
            status: Status::NEW,
            client: client,
        }
    }
}

impl Controller for Session {
    fn get_created_at(&self) -> Instant {
        self.created_at
    }

    fn get_touch_at(&self) -> Instant {
        self.touch_at
    }

    fn get_deadline(&self) -> Instant {
        self.created_at + self.timeout
    }

    fn get_status(&self) -> &Status {
        &self.status
    }

    fn get_cookie(&self) -> &str {
        &self.cookie
    }

    fn match_cookie(&self, cookie: String) -> bool {
        self.cookie == cookie
    }
}

File: model/session/session.go
package session

import (
	"time"

	"github.com/alvidir/tp-auth/model/client"
	pb "github.com/alvidir/tp-auth/proto/client"
)

// A session represents the single session for a client
type session struct {
	client.Controller
	Cookie    string    `json:"cookie" bson:"_id,omitempty"`
	CreatedAt time.Time `json:"created_at" bson:"created_at"`
	TouchAt   time.Time `json:"touch" bson:"touch"`
	Deadline  time.Time `json:"deadline" bson:"deadline"`
	Status    pb.Status `json:"status" bson:"status"`
}

func (session *session) GetCreatedAt() time.Time {
	return session.CreatedAt
}

func (session *session) GetTouchAt() time.Time {
	return session.TouchAt
}

func (session *session) GetDeadline() time.Time {
	return session.Deadline
}

func (session *session) MatchCookie(cookie string) bool {
	return session.Cookie == cookie
}

func (session *session) GetCookie() (cookie string) {
	return session.Cookie
}

func (session *session) SessionStatus() pb.Status {
	return session.Status
}
